server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsHealthIndicatorService.java

```diff
@@ -0,0 +1,251 @@
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.cluster.health.ClusterHealthStatus;
+import org.elasticsearch.cluster.health.ClusterShardHealth;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.UnassignedInfo;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.health.HealthIndicatorDetails;
+import org.elasticsearch.health.HealthIndicatorResult;
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.common.Strings.collectionToDelimitedStringWithLimit;
+import static org.elasticsearch.common.util.CollectionUtils.limitSize;
+import static org.elasticsearch.health.HealthStatus.GREEN;
+import static org.elasticsearch.health.HealthStatus.RED;
+import static org.elasticsearch.health.HealthStatus.YELLOW;
+import static org.elasticsearch.health.ServerHealthComponents.DATA;
+
+/**
+ * This indicator reports health for shards.
+ *
+ * Indicator will report:
+ * * RED when one or more shards are not available
+ * * YELLOW when one or more shards are not replicated
+ * * GREEN otherwise
+ *
+ * Each shard needs to be available and replicated in order to guarantee high availability and prevent data loses.
+ */
+public class ShardsHealthIndicatorService implements HealthIndicatorService {
+
+    public static final String NAME = "shards";
+
+    private final ClusterService clusterService;
+
+    public ShardsHealthIndicatorService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public String component() {
+        return DATA;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate() {
+
+        var state = clusterService.state();
+        var stats = new ShardAllocationStats();
+
+        for (IndexRoutingTable indexShardRouting : state.routingTable()) {
+            for (IndexShardRoutingTable shardRouting : indexShardRouting) {
+                var primaryShard = shardRouting.primaryShard();
+
+                if (primaryShard.active()) {
+                    stats.allocatedPrimaries++;
+                } else if (isInitializing(primaryShard)) {
+                    stats.initializingPrimaries.add(primaryShard.shardId());
+                } else {
+                    stats.unallocatedPrimaries.add(primaryShard.shardId());
+                }
+
+                if (shardRouting.replicaShards().isEmpty()) {
```

Comment:
I just realized I should also check if it is a shard on clod/frozen tier backed by a snapshot. In such case it should not be reported as yellow